infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java

```diff
@@ -0,0 +1,179 @@
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.String.format;
+
+import com.google.common.annotations.Beta;
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.EnvVarBuilder;
+import io.fabric8.kubernetes.api.model.EnvVarSourceBuilder;
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.LabelSelectorBuilder;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretKeySelectorBuilder;
+import io.fabric8.kubernetes.api.model.SecretVolumeSourceBuilder;
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeBuilder;
+import io.fabric8.kubernetes.api.model.VolumeMountBuilder;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+
+/**
+ * Finds secrets with specific labels in namespace, and mount their values as file or environment
+ * variable into all (or specified by "targetContainer" annotation) workspace containers.
+ * Secrets with annotation "useSecretAsEnv:true" are mount as env variables, env name is
+ * read from "envName" or "<datakey>.envName>" annotation. Secrets which don't have it, are
+ * mounted as file in the folder specified by "mountPath" annotation. Refer to che-docs for
+ * concrete examples.
+ */
+@Beta
+@Singleton
+public class SecretAsContainerResourceProvisioner<E extends KubernetesEnvironment> {
+
+  private final Map<String, String> secretLabels;
+
+  @Inject
+  public SecretAsContainerResourceProvisioner(
+      @Named("che.workspace.provision.secret.labels") String[] labels) {
+    this.secretLabels = new HashMap<>();
+    for (String keyValue : labels) {
+      String[] pair = keyValue.split(":", 2);
+      secretLabels.put(pair[0], pair.length == 1 ? "" : pair[1]);
+    }
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      boolean mountAsEnv =
+          parseBoolean(
+              secret.getMetadata().getAnnotations().getOrDefault("useSecretAsEnv", "false"));
```

Comment:
They are annotations, not labels, so as for me no much reason to make them prefixed, so lets keep writes job easier, wdyt ?

Reply:
Annotations can be random: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

Reply:
Taken from the above mentioned docs:

> If the prefix is omitted, the annotation Key is presumed to be private to the user. Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add annotations to end-user objects must specify a prefix.

So I think it is actually required of us to use the prefix. You could argue that we're just reading the annotation, not adding it as the docs suggest, but my counter-argument would be that we are the sole users of that annotation and that we should not "invade" the user's "namespace".

There is also a prior art to this style of annotation declarations - the ingress annotations that are also defined by the user, yet they are prefixed, because they are specific to certain ingress controllers.

Reply:
Prefixes was already agreed with Mario and applied.